### PR TITLE
Further improved order of auto-complete results (amends #469).

### PR DESCRIPTION
**Pull request by IonoclastBrigham**
Ok, so here's the rundown: 

1. I build a second pattern that is used to find the length of the longest prefix substring match.
2. I find the length of matching prefixes using `string.find()`; if one is longer than the other, the *longer* match always sorts first.
3. If prefix matches are equal, it falls back on the total original pattern match to find the substring with the fewest letters between matches; if they are equal collate, alphanumerically.
4. Else, the *shorter* match sorts first.

This patch needs review on a number of points:

* **Test coverage and corner cases** – It passes the relevant test, which I modified to match my expected output, but it's not exceedingly thorough. It does give expected results for all the completions I could think to type into the editor, though.
* **Coding style** – I tried to match the style I found in the surrounding code, but there doesn't seem to be a consistent standard for things like spacing between punctuation and use of parentheses. At least, not that I could infer by looking a the code.
* **Efficiency** – This might not by a huge concern, since this this code will only run at most a few times a second, but if there is any obvious wastage or simple changes that could be made, it may be worth doing. For example, I switched from `string.gsub()` to `string.find()` for calculating pattern match lengths because it avoids allocating a closure and is arguably clearer as to intent, but it does return unused `number` and `string` results. Or, the `string.sub()` operation to chop the prefix off of `pat` could be memoized in a cache `table` to avoid redundant string operations and gc churn.
* **Magic number** – I explicitly use the number `13` as a literal in the above substring operation. It is called out and explained in a comment. It could be assigned to a named "constant" somewhere, but this doesn't really give any more information or future-proofing than the comment does. It would be better to calculate this value at runtime from the pattern, but the best way to implement this isn't immediately obvious to me. Possibly using a named format string when building `pat`, instead of manually concatenating values; we could get the length of that string at runtime.

And finally, *is this actually what most everyone probably wants and expects*, or is it just a quirk of my brain that this solution is preferable? This is perfectly possible.

### CODE REVIEW

**Comment by pkulchenko** [comment]
> is this actually what most everyone probably wants and expects, or is it just a quirk of my brain that this solution is preferable?

@IonoclastBrigham, thank you for the patch and detailed description. I'll answer your last question first; I think this is desirable behavior and I've tried different combinations and it does seem to provide a better alternatives than the current mechanism (or at least I couldn't find a counter examples).

I did come up with a slightly different approach, mostly because I wanted to avoid "magic" numbers (I do have a "penalty" constant, but it's of a different nature ;) ) and also wanted to extract the processing from the sort, as it's done for each comparison. Here is the diff for the logic I have; it should satisfy the criteria we discussed in the other ticket:

```diff
diff --git a/src/editor/autocomplete.lua b/src/editor/autocomplete.lua
index 6311c89..9355d41 100644
--- a/src/editor/autocomplete.lua
+++ b/src/editor/autocomplete.lua
@@ -618,25 +618,24 @@ function CreateAutoCompList(editor,key,pos)
       local strategy = ide.config.acandtip.strategy
 
       if (strategy == 2 and #apilist < 128) then
-        local num, max = 0, #rest
-        local pat = rest:gsub(".",function(c)
-            num = num + 1
-            local l = c:lower()..c:upper()
-            -- don't include trailing characters
-            return "["..l.."]" .. (num < max and "([^"..l.." ]*)" or "")
-          end)
-
-        table.sort(apilist,function(a,b)
-            local ma,mb = 0,0
-            a:gsub(pat,function(...)
-                local l = {...}
-                for _, v in ipairs(l) do ma = ma + #v end
-              end)
-            b:gsub(pat,function(...)
+        -- when matching "ret": "ret." < "re.t" < "r.et"
+        local pat = rest:gsub(".", function(c) return "["..c:lower()..c:upper().."](.-)" end)
+        local weights = {}
+        local penalty = 0.1
+        local function weight(str)
+          if not weights[str] then
+            local w = 0
+            str:gsub(pat,function(...)
                 local l = {...}
-                for _, v in ipairs(l) do mb = mb + #v end
+                -- penalize gaps between matches, more so at the beginning
+                for n, v in ipairs(l) do w = w + #v * (1 + (#l-n)*penalty) end
               end)
-
+            weights[str] = w
+          end
+          return weights[str]
+        end
+        table.sort(apilist,function(a,b)
+            local ma, mb = weight(a), weight(b)
             if (ma == mb) then return a:lower()<b:lower() end
             return ma<mb
           end)
```

**Comment by pkulchenko** [comment]
@IonoclastBrigham, does the patched version work for you?

**Comment by IonoclastBrigham** [comment]
Yeah, when I revert `autocomplete.lua` to `7aa89ed` and apply the patch, it looks like it works the same on all the inputs I typed in. Had a little trouble getting the patch applied for some reason, but it looks good now.

I actually like this solution better. Much more elegant. Still feels weird to use `string.gsub()` to iterate over chars in a string, but apparently [this is accepted practice in the Lua community](http://stackoverflow.com/questions/829063/how-to-iterate-individual-characters-in-lua-string/832414#832414).

I revised this pull request with your changes, so you get the updated test for free if you just merge.

**Comment by pkulchenko** [comment]
@IonoclastBrigham, it seems like the autocomplete changes look good, but for some reason it includes the three other commits (probably because of the revert) as well as the original commit. If you look at the "Files changed" tab, you will see 6 files changed instead of only 2.

What I usually do when submit patches is to create a separate branch, which makes it easier for those who pull the changes in to take them and merge or modify if needed; it also allows you to simply drop the branch and pull the changes into master without any merge commits (if there are any changes).

Could you please fork it again and re-apply the patch so that it applies cleanly to the current master without other commits? I'd much appreciate it. Thank you.

**Comment by IonoclastBrigham** [comment]
I'll see about creating a new pull request.
